From: Rene Kita <mail@rkta.de>
Date: Tue Sep 17 11:52:48 CEST 2024
Subject: Use Strnulterm everywhere
Patch: patches/Use-Strnulterm-everywhere.patch

Use a macro to nul-terminated Str. Macro exists but is mostly unused.

While at it, refactor Strshrink.
---
 Str.c  |   46 +++++++++++++++++++++++++++-------------------
 Str.h  |    1 -
 etc.c  |    2 +-
 file.c |    3 +--
 4 files changed, 29 insertions(+), 23 deletions(-)

--- a/Str.c
+++ b/Str.c
@@ -26,6 +26,13 @@
 
 #define INITIAL_STR_SIZE 32
 
+static Str
+Strnulterm(Str x)
+{
+    x->ptr[x->length] = '\0';
+    return x;
+}
+
 Str
 Strnew(void)
 {
@@ -35,9 +42,9 @@ Strnew(void)
     x->ptr = GC_MALLOC_ATOMIC(INITIAL_STR_SIZE);
     if (x->ptr == NULL)
 	exit(1);
-    x->ptr[0] = '\0';
     x->area_size = INITIAL_STR_SIZE;
     x->length = 0;
+    Strnulterm(x);
     return x;
 }
 
@@ -54,9 +61,9 @@ Strnew_size(int n)
     x->ptr = GC_MALLOC_ATOMIC(n + 1);
     if (x->ptr == NULL)
 	exit(1);
-    x->ptr[0] = '\0';
     x->area_size = n + 1;
     x->length = 0;
+    Strnulterm(x);
     return x;
 }
 
@@ -83,7 +90,7 @@ Strnew_charp(const char *p)
     x->area_size = n;
     x->length = len;
     memmove(x->ptr, p, len);
-    x->ptr[x->length] = '\0';
+    Strnulterm(x);
     return x;
 }
 
@@ -124,7 +131,7 @@ Strnew_charp_n(const char *p, int n)
     x->area_size = n + 1;
     x->length = len;
     memmove(x->ptr, p, len);
-    x->ptr[x->length] = '\0';
+    Strnulterm(x);
     return x;
 }
 
@@ -140,7 +147,7 @@ void
 Strclear(Str s)
 {
     s->length = 0;
-    s->ptr[0] = '\0';
+    Strnulterm(s);
 }
 
 void
@@ -170,7 +177,7 @@ Strcopy_charp(Str x, const char *y)
 
     if (y == NULL) {
 	x->length = 0;
-	x->ptr[0] = '\0';
+	Strnulterm(x);
 	return;
     }
     len = strlen(y);
@@ -183,8 +190,8 @@ Strcopy_charp(Str x, const char *y)
 	x->area_size = len + 1;
     }
     memmove(x->ptr, y, len);
-    x->ptr[len] = '\0';
     x->length = len;
+    Strnulterm(x);
 }
 
 void
@@ -194,7 +201,7 @@ Strcopy_charp_n(Str x, const char *y, in
 
     if (y == NULL) {
 	x->length = 0;
-	x->ptr[0] = '\0';
+	Strnulterm(x);
 	return;
     }
     if (len < 0 || len >= STR_SIZE_MAX)
@@ -206,8 +213,8 @@ Strcopy_charp_n(Str x, const char *y, in
 	x->area_size = len + 1;
     }
     memmove(x->ptr, y, len);
-    x->ptr[len] = '\0';
     x->length = len;
+    Strnulterm(x);
 }
 
 void
@@ -237,7 +244,7 @@ Strcat_charp_n(Str x, const char *y, int
     }
     memmove(&x->ptr[x->length], y, n);
     x->length += n;
-    x->ptr[x->length] = '\0';
+    Strnulterm(x);
 }
 
 void
@@ -289,7 +296,7 @@ Strgrow(Str x)
 	    exit(1);
 	x->area_size = newlen;
     }
-    x->ptr[x->length] = '\0';
+    Strnulterm(x);
 }
 
 Str
@@ -329,7 +336,7 @@ Strchop(Str s)
 	   (s->ptr[s->length - 1] == '\n' || s->ptr[s->length - 1] == '\r')) {
 	s->length--;
     }
-    s->ptr[s->length] = '\0';
+    Strnulterm(s);
 }
 
 void
@@ -344,7 +351,8 @@ Strinsert_char(Str s, int pos, char c)
 	return;
     for (i = s->length; i > pos; i--)
 	s->ptr[i] = s->ptr[i - 1];
-    s->ptr[++s->length] = '\0';
+    s->length++;
+    Strnulterm(s);
     s->ptr[pos] = c;
 }
 
@@ -364,14 +372,14 @@ Strdelete(Str s, int pos, int n)
     if (n < 0)
 	n = STR_SIZE_MAX - pos - 1;
     if (s->length <= pos + n) {
-	s->ptr[pos] = '\0';
 	s->length = pos;
+	Strnulterm(s);
 	return;
     }
     for (i = pos; i < s->length - n; i++)
 	s->ptr[i] = s->ptr[i + n];
-    s->ptr[i] = '\0';
     s->length = i;
+    Strnulterm(s);
 }
 
 void
@@ -379,8 +387,8 @@ Strtruncate(Str s, int pos)
 {
     if (pos < 0 || s->length < pos)
 	return;
-    s->ptr[pos] = '\0';
     s->length = pos;
+    Strnulterm(s);
 }
 
 void
@@ -388,11 +396,11 @@ Strshrink(Str s, int n)
 {
     if (n >= s->length) {
 	s->length = 0;
-	s->ptr[0] = '\0';
+	Strnulterm(s);
     }
     else if (n > 0) {
 	s->length -= n;
-	s->ptr[s->length] = '\0';
+	Strnulterm(s);
     }
 }
 
@@ -414,7 +422,7 @@ Strremovetrailingspaces(Str s)
 
     for (i = s->length - 1; i >= 0 && IS_SPACE(s->ptr[i]); i--) ;
     s->length = i + 1;
-    s->ptr[i + 1] = '\0';
+    Strnulterm(s);
 }
 
 Str
--- a/file.c
+++ b/file.c
@@ -4521,8 +4521,7 @@ getMetaRefreshParam(const char *q, Str *
 	    if (s_tmp->length > 0 &&
 	        (s_tmp->ptr[s_tmp->length - 1] == '\"' ||	/* " */
 		 s_tmp->ptr[s_tmp->length - 1] == '\'')) {	/* ' */
-		s_tmp->length--;
-		s_tmp->ptr[s_tmp->length] = '\0';
+		Strshrink(s_tmp, 1);
 	    }
 	    q = r;
 	}
--- a/Str.h
+++ b/Str.h
@@ -68,7 +68,6 @@ void Strgrow(Str s);
 
 #define STR_SIZE_MAX (INT_MAX / 32)
 #define Strcat_char(x,y) (((x)->length+1>=STR_SIZE_MAX)?0:(((x)->length+1>=(x)->area_size)?Strgrow(x),0:0,(x)->ptr[(x)->length++]=(y),(x)->ptr[(x)->length]=0))
-#define Strnulterm(x) ((x)->ptr[(x)->length]=0)
 #define Strcmp(x,y)                  strcmp((x)->ptr,(y)->ptr)
 #define Strcmp_charp(x,y)            strcmp((x)->ptr,(y))
 #define Strcasecmp(x,y)              strcasecmp((x)->ptr,(y)->ptr)
--- a/etc.c
+++ b/etc.c
@@ -2159,7 +2159,7 @@ base64_encode(const char *src, size_t le
 	}
 	Strcatc(dest, '=');
     }
-    Strnulterm(dest);
+    dest->ptr[dest->length]='\0';
     return dest;
 #undef Strcatc
 }
